File: 48_positional_encoding_timeseries/python/strategy.py
"""
Trading Strategy and Backtesting

This module provides trading strategy implementation and backtesting
utilities for evaluating time series models.
"""

import numpy as np
import pandas as pd
from typing import Dict, List, Optional, Tuple
from dataclasses import dataclass
from enum import Enum


class SignalType(Enum):
    """Trading signal types."""
    LONG = 1
    NEUTRAL = 0
    SHORT = -1


@dataclass
class BacktestResult:
    """
    Container for backtest results.

    Attributes:
        returns: Daily/period returns
        cumulative_returns: Cumulative returns
        equity_curve: Portfolio value over time
        sharpe_ratio: Annualized Sharpe ratio
        sortino_ratio: Sortino ratio
        max_drawdown: Maximum drawdown
        total_return: Total period return
        win_rate: Percentage of winning trades
        n_trades: Number of trades
        metrics: Additional metrics dictionary
    """
    returns: np.ndarray
    cumulative_returns: np.ndarray
    equity_curve: np.ndarray
    sharpe_ratio: float
    sortino_ratio: float
    max_drawdown: float
    total_return: float
    win_rate: float
    n_trades: int
    metrics: Dict


class TradingStrategy:
    """
    Base trading strategy class.

    Converts model predictions to trading signals and manages positions.

    Args:
        threshold: Signal threshold for position entry
        max_position: Maximum position size (1.0 = 100%)
        transaction_cost: Transaction cost per trade (fraction)
        stop_loss: Stop loss threshold (fraction)
        take_profit: Take profit threshold (fraction)

    Example:
        >>> strategy = TradingStrategy(threshold=0.001)
        >>> signals = strategy.generate_signals(predictions, threshold=0.001)
    """

    def __init__(
        self,
        threshold: float = 0.001,
        max_position: float = 1.0,
        transaction_cost: float = 0.001,
        stop_loss: Optional[float] = None,
        take_profit: Optional[float] = None
    ):
        self.threshold = threshold
        self.max_position = max_position
        self.transaction_cost = transaction_cost
        self.stop_loss = stop_loss
        self.take_profit = take_profit

    def generate_signals(
        self,
        predictions: np.ndarray,
        threshold: Optional[float] = None
    ) -> np.ndarray:
        """
        Generate trading signals from predictions.

        Args:
            predictions: Model predictions (returns or directions)
            threshold: Override default threshold

        Returns:
            Array of signals (-1, 0, 1)
        """
        if threshold is None:
            threshold = self.threshold

        signals = np.zeros_like(predictions)
        signals[predictions > threshold] = 1
        signals[predictions < -threshold] = -1

        return signals

    def generate_position_sizes(
        self,
        signals: np.ndarray,
        confidences: Optional[np.ndarray] = None
    ) -> np.ndarray:
        """
        Generate position sizes from signals.

        Args:
            signals: Trading signals (-1, 0, 1)
            confidences: Optional confidence scores

        Returns:
            Position sizes
        """
        if confidences is not None:
            # Scale position by confidence
            positions = signals * confidences * self.max_position
        else:
            positions = signals * self.max_position

        return np.clip(positions, -self.max_position, self.max_position)


def calculate_metrics(
    returns: np.ndarray,
    periods_per_year: int = 252
) -> Dict[str, float]:
    """
    Calculate trading performance metrics.

    Args:
        returns: Array of period returns
        periods_per_year: Number of periods per year (252 for daily, 8760 for hourly)

    Returns:
        Dictionary of performance metrics
    """
    # Filter out NaN values
    returns = returns[~np.isnan(returns)]

    if len(returns) == 0:
        return {
            'total_return': 0.0,
            'annual_return': 0.0,
            'volatility': 0.0,
            'sharpe_ratio': 0.0,
            'sortino_ratio': 0.0,
            'max_drawdown': 0.0,
            'calmar_ratio': 0.0,
            'win_rate': 0.0
        }

    # Total return
    cumulative = np.cumprod(1 + returns)
    total_return = cumulative[-1] - 1

    # Annualized return
    n_periods = len(returns)
    annual_return = (1 + total_return) ** (periods_per_year / n_periods) - 1

    # Volatility
    volatility = np.std(returns) * np.sqrt(periods_per_year)

    # Sharpe ratio (assuming 0 risk-free rate)
    if volatility > 0:
        sharpe_ratio = annual_return / volatility
    else:
        sharpe_ratio = 0.0

    # Sortino ratio (downside volatility)
    downside_returns = returns[returns < 0]
    if len(downside_returns) > 0:
        downside_vol = np.std(downside_returns) * np.sqrt(periods_per_year)
        sortino_ratio = annual_return / downside_vol if downside_vol > 0 else 0.0
    else:
        sortino_ratio = sharpe_ratio * 1.5  # Approximate if no downside

    # Maximum drawdown
    cumulative = np.cumprod(1 + returns)
    running_max = np.maximum.accumulate(cumulative)
    drawdowns = cumulative / running_max - 1
    max_drawdown = np.min(drawdowns)

    # Calmar ratio
    calmar_ratio = annual_return / abs(max_drawdown) if max_drawdown != 0 else 0.0

    # Win rate
    winning_trades = np.sum(returns > 0)
    total_trades = np.sum(returns != 0)
    win_rate = winning_trades / total_trades if total_trades > 0 else 0.0

    return {
        'total_return': total_return,
        'annual_return': annual_return,
        'volatility': volatility,
        'sharpe_ratio': sharpe_ratio,
        'sortino_ratio': sortino_ratio,
        'max_drawdown': max_drawdown,
        'calmar_ratio': calmar_ratio,
        'win_rate': win_rate
    }


def run_backtest(
    predictions: np.ndarray,
    actual_returns: np.ndarray,
    strategy: Optional[TradingStrategy] = None,
    initial_capital: float = 100000,
    periods_per_year: int = 8760  # Hourly data
) -> BacktestResult:
    """
    Run backtest on model predictions.

    Args:
        predictions: Model return predictions
        actual_returns: Actual market returns
        strategy: Trading strategy (uses default if None)
        initial_capital: Starting capital
        periods_per_year: Periods per year for annualization

    Returns:
        BacktestResult with performance metrics

    Example:
        >>> predictions = model(test_data)['predictions'].numpy()
        >>> result = run_backtest(predictions, actual_returns)
        >>> print(f"Sharpe: {result.sharpe_ratio:.2f}")
    """
    if strategy is None:
        strategy = TradingStrategy()

    # Ensure arrays are 1D
    if predictions.ndim > 1:
        predictions = predictions[:, 0]
    if actual_returns.ndim > 1:
        actual_returns = actual_returns[:, 0]

    # Align lengths
    min_len = min(len(predictions), len(actual_returns))
    predictions = predictions[:min_len]
    actual_returns = actual_returns[:min_len]

    # Generate signals
    signals = strategy.generate_signals(predictions)
    positions = strategy.generate_position_sizes(signals)

    # Calculate strategy returns
    strategy_returns = positions * actual_returns

    # Apply transaction costs
    position_changes = np.abs(np.diff(positions, prepend=0))
    transaction_costs = position_changes * strategy.transaction_cost
    strategy_returns = strategy_returns - transaction_costs

    # Calculate equity curve
    equity_curve = initial_capital * np.cumprod(1 + strategy_returns)

    # Calculate metrics
    metrics = calculate_metrics(strategy_returns, periods_per_year)

    # Cumulative returns
    cumulative_returns = np.cumprod(1 + strategy_returns) - 1

    # Count trades
    n_trades = np.sum(position_changes > 0)

    return BacktestResult(
        returns=strategy_returns,
        cumulative_returns=cumulative_returns,
        equity_curve=equity_curve,
        sharpe_ratio=metrics['sharpe_ratio'],
        sortino_ratio=metrics['sortino_ratio'],
        max_drawdown=metrics['max_drawdown'],
        total_return=metrics['total_return'],
        win_rate=metrics['win_rate'],
        n_trades=n_trades,
        metrics=metrics
    )


def compare_strategies(
    predictions: np.ndarray,
    actual_returns: np.ndarray,
    thresholds: List[float] = [0.0005, 0.001, 0.002],
    periods_per_year: int = 8760
) -> pd.DataFrame:
    """
    Compare different strategy thresholds.

    Args:
        predictions: Model predictions
        actual_returns: Actual returns
        thresholds: List of thresholds to test
        periods_per_year: Periods per year

    Returns:
        DataFrame comparing strategy performance
    """
    results = []

    for threshold in thresholds:
        strategy = TradingStrategy(threshold=threshold)
        result = run_backtest(
            predictions, actual_returns, strategy,
            periods_per_year=periods_per_year
        )

        results.append({
            'threshold': threshold,
            'total_return': result.total_return,
            'sharpe_ratio': result.sharpe_ratio,
            'sortino_ratio': result.sortino_ratio,
            'max_drawdown': result.max_drawdown,
            'win_rate': result.win_rate,
            'n_trades': result.n_trades
        })

    return pd.DataFrame(results)


def calculate_buy_and_hold(
    actual_returns: np.ndarray,
    initial_capital: float = 100000,
    periods_per_year: int = 8760
) -> BacktestResult:
    """
    Calculate buy and hold benchmark.

    Args:
        actual_returns: Market returns
        initial_capital: Starting capital
        periods_per_year: Periods per year

    Returns:
        BacktestResult for buy and hold strategy
    """
    if actual_returns.ndim > 1:
        actual_returns = actual_returns[:, 0]

    equity_curve = initial_capital * np.cumprod(1 + actual_returns)
    cumulative_returns = np.cumprod(1 + actual_returns) - 1
    metrics = calculate_metrics(actual_returns, periods_per_year)

    return BacktestResult(
        returns=actual_returns,
        cumulative_returns=cumulative_returns,
        equity_curve=equity_curve,
        sharpe_ratio=metrics['sharpe_ratio'],
        sortino_ratio=metrics['sortino_ratio'],
        max_drawdown=metrics['max_drawdown'],
        total_return=metrics['total_return'],
        win_rate=metrics['win_rate'],
        n_trades=1,
        metrics=metrics
    )


if __name__ == "__main__":
    # Test strategy and backtesting
    print("Testing Trading Strategy and Backtesting...")
    print("=" * 60)

    # Generate synthetic predictions and returns
    np.random.seed(42)
    n_samples = 1000

    # Simulate model predictions with some signal
    true_signal = np.random.normal(0, 0.01, n_samples)
    predictions = true_signal + np.random.normal(0, 0.005, n_samples)
    actual_returns = true_signal + np.random.normal(0, 0.02, n_samples)

    # Test strategy
    print("\n1. Testing TradingStrategy...")
    strategy = TradingStrategy(threshold=0.001, transaction_cost=0.001)
    signals = strategy.generate_signals(predictions)
    positions = strategy.generate_position_sizes(signals)
    print(f"   Long signals: {np.sum(signals > 0)}")
    print(f"   Short signals: {np.sum(signals < 0)}")
    print(f"   Neutral: {np.sum(signals == 0)}")

    # Test backtest
    print("\n2. Running backtest...")
    result = run_backtest(predictions, actual_returns, strategy)
    print(f"   Total Return: {result.total_return:.2%}")
    print(f"   Sharpe Ratio: {result.sharpe_ratio:.2f}")
    print(f"   Sortino Ratio: {result.sortino_ratio:.2f}")
    print(f"   Max Drawdown: {result.max_drawdown:.2%}")
    print(f"   Win Rate: {result.win_rate:.2%}")
    print(f"   Trades: {result.n_trades}")

    # Compare with buy and hold
    print("\n3. Buy and Hold benchmark...")
    bh_result = calculate_buy_and_hold(actual_returns)
    print(f"   Total Return: {bh_result.total_return:.2%}")
    print(f"   Sharpe Ratio: {bh_result.sharpe_ratio:.2f}")

    # Compare thresholds
    print("\n4. Comparing thresholds...")
    comparison = compare_strategies(predictions, actual_returns)
    print(comparison.to_string(index=False))

    print("\n" + "=" * 60)
    print("Strategy tests completed!")

File: 48_positional_encoding_timeseries/rust/src/strategy.rs
//! Trading Strategy and Backtesting
//!
//! This module provides trading strategy implementations and backtesting
//! utilities for evaluating time series models.

use std::f64;

/// Trading signal types
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Signal {
    Long,
    Neutral,
    Short,
}

impl Signal {
    /// Convert to numeric value
    pub fn as_f64(&self) -> f64 {
        match self {
            Signal::Long => 1.0,
            Signal::Neutral => 0.0,
            Signal::Short => -1.0,
        }
    }

    /// Create from prediction value
    pub fn from_prediction(pred: f64, threshold: f64) -> Self {
        if pred > threshold {
            Signal::Long
        } else if pred < -threshold {
            Signal::Short
        } else {
            Signal::Neutral
        }
    }
}

/// Container for backtest results
#[derive(Debug, Clone)]
pub struct BacktestResult {
    /// Daily/period returns
    pub returns: Vec<f64>,
    /// Cumulative returns
    pub cumulative_returns: Vec<f64>,
    /// Portfolio value over time
    pub equity_curve: Vec<f64>,
    /// Annualized Sharpe ratio
    pub sharpe_ratio: f64,
    /// Sortino ratio
    pub sortino_ratio: f64,
    /// Maximum drawdown
    pub max_drawdown: f64,
    /// Total period return
    pub total_return: f64,
    /// Percentage of winning trades
    pub win_rate: f64,
    /// Number of trades
    pub n_trades: usize,
    /// Additional metrics
    pub metrics: PerformanceMetrics,
}

/// Performance metrics
#[derive(Debug, Clone)]
pub struct PerformanceMetrics {
    pub annual_return: f64,
    pub volatility: f64,
    pub calmar_ratio: f64,
}

/// Trading strategy configuration
#[derive(Debug, Clone)]
pub struct TradingStrategy {
    /// Signal threshold for position entry
    pub threshold: f64,
    /// Maximum position size (1.0 = 100%)
    pub max_position: f64,
    /// Transaction cost per trade (fraction)
    pub transaction_cost: f64,
    /// Stop loss threshold (fraction)
    pub stop_loss: Option<f64>,
    /// Take profit threshold (fraction)
    pub take_profit: Option<f64>,
}

impl Default for TradingStrategy {
    fn default() -> Self {
        Self {
            threshold: 0.001,
            max_position: 1.0,
            transaction_cost: 0.001,
            stop_loss: None,
            take_profit: None,
        }
    }
}

impl TradingStrategy {
    /// Create a new trading strategy
    pub fn new(threshold: f64) -> Self {
        Self {
            threshold,
            ..Default::default()
        }
    }

    /// Set maximum position size
    pub fn with_max_position(mut self, max_position: f64) -> Self {
        self.max_position = max_position;
        self
    }

    /// Set transaction cost
    pub fn with_transaction_cost(mut self, cost: f64) -> Self {
        self.transaction_cost = cost;
        self
    }

    /// Set stop loss
    pub fn with_stop_loss(mut self, stop_loss: f64) -> Self {
        self.stop_loss = Some(stop_loss);
        self
    }

    /// Set take profit
    pub fn with_take_profit(mut self, take_profit: f64) -> Self {
        self.take_profit = Some(take_profit);
        self
    }

    /// Generate trading signals from predictions
    pub fn generate_signals(&self, predictions: &[f64]) -> Vec<Signal> {
        predictions
            .iter()
            .map(|&p| Signal::from_prediction(p, self.threshold))
            .collect()
    }

    /// Generate position sizes from signals
    pub fn generate_positions(&self, signals: &[Signal], confidences: Option<&[f64]>) -> Vec<f64> {
        match confidences {
            Some(conf) => signals
                .iter()
                .zip(conf.iter())
                .map(|(s, c)| (s.as_f64() * c * self.max_position).clamp(-self.max_position, self.max_position))
                .collect(),
            None => signals
                .iter()
                .map(|s| s.as_f64() * self.max_position)
                .collect(),
        }
    }
}

/// Calculate trading performance metrics
pub fn calculate_metrics(returns: &[f64], periods_per_year: usize) -> PerformanceMetrics {
    // Filter NaN values
    let returns: Vec<f64> = returns.iter().copied().filter(|r| r.is_finite()).collect();

    if returns.is_empty() {
        return PerformanceMetrics {
            annual_return: 0.0,
            volatility: 0.0,
            calmar_ratio: 0.0,
        };
    }

    // Total return
    let cumulative: f64 = returns.iter().map(|r| 1.0 + r).product();
    let total_return = cumulative - 1.0;

    // Annualized return
    let n_periods = returns.len() as f64;
    let annual_return =
        (1.0 + total_return).powf(periods_per_year as f64 / n_periods) - 1.0;

    // Volatility
    let mean = returns.iter().sum::<f64>() / n_periods;
    let variance = returns.iter().map(|r| (r - mean).powi(2)).sum::<f64>() / n_periods;
    let volatility = variance.sqrt() * (periods_per_year as f64).sqrt();

    // Maximum drawdown
    let mut peak = 1.0;
    let mut max_dd = 0.0;
    let mut equity = 1.0;
    for r in &returns {
        equity *= 1.0 + r;
        if equity > peak {
            peak = equity;
        }
        let dd = (peak - equity) / peak;
        if dd > max_dd {
            max_dd = dd;
        }
    }

    // Calmar ratio
    let calmar_ratio = if max_dd > 0.0 {
        annual_return / max_dd
    } else {
        0.0
    };

    PerformanceMetrics {
        annual_return,
        volatility,
        calmar_ratio,
    }
}

/// Run backtest on model predictions
pub fn run_backtest(
    predictions: &[f64],
    actual_returns: &[f64],
    strategy: &TradingStrategy,
    initial_capital: f64,
    periods_per_year: usize,
) -> BacktestResult {
    let min_len = predictions.len().min(actual_returns.len());
    let predictions = &predictions[..min_len];
    let actual_returns = &actual_returns[..min_len];

    // Generate signals and positions
    let signals = strategy.generate_signals(predictions);
    let positions = strategy.generate_positions(&signals, None);

    // Calculate strategy returns
    let mut strategy_returns = Vec::with_capacity(min_len);
    let mut position_changes = Vec::with_capacity(min_len);

    for i in 0..min_len {
        let prev_position = if i > 0 { positions[i - 1] } else { 0.0 };
        let position_change = (positions[i] - prev_position).abs();
        position_changes.push(position_change);

        let gross_return = positions[i] * actual_returns[i];
        let transaction_cost = position_change * strategy.transaction_cost;
        strategy_returns.push(gross_return - transaction_cost);
    }

    // Calculate equity curve
    let mut equity = initial_capital;
    let mut equity_curve = Vec::with_capacity(min_len);
    for &r in &strategy_returns {
        equity *= 1.0 + r;
        equity_curve.push(equity);
    }

    // Calculate cumulative returns
    let mut cumulative = 1.0;
    let cumulative_returns: Vec<f64> = strategy_returns
        .iter()
        .map(|&r| {
            cumulative *= 1.0 + r;
            cumulative - 1.0
        })
        .collect();

    // Calculate metrics
    let metrics = calculate_metrics(&strategy_returns, periods_per_year);

    // Sharpe ratio
    let mean_return = strategy_returns.iter().sum::<f64>() / strategy_returns.len() as f64;
    let std_return = (strategy_returns
        .iter()
        .map(|r| (r - mean_return).powi(2))
        .sum::<f64>()
        / strategy_returns.len() as f64)
        .sqrt();
    let sharpe_ratio = if std_return > 0.0 {
        metrics.annual_return / (std_return * (periods_per_year as f64).sqrt())
    } else {
        0.0
    };

    // Sortino ratio (downside volatility)
    let downside_returns: Vec<f64> = strategy_returns
        .iter()
        .copied()
        .filter(|&r| r < 0.0)
        .collect();
    let sortino_ratio = if !downside_returns.is_empty() {
        let downside_vol = (downside_returns
            .iter()
            .map(|r| r.powi(2))
            .sum::<f64>()
            / downside_returns.len() as f64)
            .sqrt()
            * (periods_per_year as f64).sqrt();
        if downside_vol > 0.0 {
            metrics.annual_return / downside_vol
        } else {
            0.0
        }
    } else {
        sharpe_ratio * 1.5 // Approximate if no downside
    };

    // Maximum drawdown
    let mut peak = initial_capital;
    let mut max_drawdown = 0.0;
    for &equity_val in &equity_curve {
        if equity_val > peak {
            peak = equity_val;
        }
        let dd = (peak - equity_val) / peak;
        if dd > max_drawdown {
            max_drawdown = dd;
        }
    }

    // Win rate
    let winning_trades = strategy_returns.iter().filter(|&&r| r > 0.0).count();
    let total_trades = strategy_returns.iter().filter(|&&r| r != 0.0).count();
    let win_rate = if total_trades > 0 {
        winning_trades as f64 / total_trades as f64
    } else {
        0.0
    };

    // Number of trades
    let n_trades = position_changes.iter().filter(|&&c| c > 0.0).count();

    // Calculate total return before moving cumulative_returns
    let total_return = cumulative_returns.last().copied().unwrap_or(0.0);

    BacktestResult {
        returns: strategy_returns,
        cumulative_returns,
        equity_curve,
        sharpe_ratio,
        sortino_ratio,
        max_drawdown,
        total_return,
        win_rate,
        n_trades,
        metrics,
    }
}

/// Calculate buy and hold benchmark
pub fn calculate_buy_and_hold(
    actual_returns: &[f64],
    initial_capital: f64,
    periods_per_year: usize,
) -> BacktestResult {
    let mut equity = initial_capital;
    let mut equity_curve = Vec::with_capacity(actual_returns.len());
    let mut cumulative = 1.0;
    let mut cumulative_returns = Vec::with_capacity(actual_returns.len());

    for &r in actual_returns {
        equity *= 1.0 + r;
        equity_curve.push(equity);
        cumulative *= 1.0 + r;
        cumulative_returns.push(cumulative - 1.0);
    }

    let metrics = calculate_metrics(actual_returns, periods_per_year);

    // Sharpe ratio
    let mean_return = actual_returns.iter().sum::<f64>() / actual_returns.len() as f64;
    let std_return = (actual_returns
        .iter()
        .map(|r| (r - mean_return).powi(2))
        .sum::<f64>()
        / actual_returns.len() as f64)
        .sqrt();
    let sharpe_ratio = if std_return > 0.0 {
        metrics.annual_return / (std_return * (periods_per_year as f64).sqrt())
    } else {
        0.0
    };

    // Maximum drawdown
    let mut peak = initial_capital;
    let mut max_drawdown = 0.0;
    for &equity_val in &equity_curve {
        if equity_val > peak {
            peak = equity_val;
        }
        let dd = (peak - equity_val) / peak;
        if dd > max_drawdown {
            max_drawdown = dd;
        }
    }

    // Win rate
    let winning = actual_returns.iter().filter(|&&r| r > 0.0).count();
    let total = actual_returns.iter().filter(|&&r| r != 0.0).count();
    let win_rate = if total > 0 {
        winning as f64 / total as f64
    } else {
        0.0
    };

    // Calculate total return before moving cumulative_returns
    let total_return = cumulative_returns.last().copied().unwrap_or(0.0);

    BacktestResult {
        returns: actual_returns.to_vec(),
        cumulative_returns,
        equity_curve,
        sharpe_ratio,
        sortino_ratio: sharpe_ratio * 1.2,
        max_drawdown,
        total_return,
        win_rate,
        n_trades: 1,
        metrics,
    }
}

/// Compare different strategy thresholds
pub fn compare_strategies(
    predictions: &[f64],
    actual_returns: &[f64],
    thresholds: &[f64],
    periods_per_year: usize,
) -> Vec<(f64, BacktestResult)> {
    thresholds
        .iter()
        .map(|&threshold| {
            let strategy = TradingStrategy::new(threshold);
            let result = run_backtest(predictions, actual_returns, &strategy, 100000.0, periods_per_year);
            (threshold, result)
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_signal_from_prediction() {
        assert_eq!(Signal::from_prediction(0.005, 0.001), Signal::Long);
        assert_eq!(Signal::from_prediction(-0.005, 0.001), Signal::Short);
        assert_eq!(Signal::from_prediction(0.0005, 0.001), Signal::Neutral);
    }

    #[test]
    fn test_strategy_generate_signals() {
        let strategy = TradingStrategy::new(0.001);
        let predictions = vec![0.005, -0.005, 0.0001, 0.002];
        let signals = strategy.generate_signals(&predictions);

        assert_eq!(signals[0], Signal::Long);
        assert_eq!(signals[1], Signal::Short);
        assert_eq!(signals[2], Signal::Neutral);
        assert_eq!(signals[3], Signal::Long);
    }

    #[test]
    fn test_strategy_generate_positions() {
        let strategy = TradingStrategy::new(0.001).with_max_position(0.5);
        let signals = vec![Signal::Long, Signal::Short, Signal::Neutral];
        let positions = strategy.generate_positions(&signals, None);

        assert!((positions[0] - 0.5).abs() < 1e-10);
        assert!((positions[1] - (-0.5)).abs() < 1e-10);
        assert!((positions[2] - 0.0).abs() < 1e-10);
    }

    #[test]
    fn test_calculate_metrics() {
        let returns = vec![0.01, -0.005, 0.02, -0.01, 0.015];
        let metrics = calculate_metrics(&returns, 252);

        assert!(metrics.annual_return.is_finite());
        assert!(metrics.volatility >= 0.0);
    }

    #[test]
    fn test_run_backtest() {
        use rand::{Rng, SeedableRng};
        use rand::rngs::StdRng;

        let mut rng = StdRng::seed_from_u64(42);
        let n = 100;

        let actual_returns: Vec<f64> = (0..n).map(|_| rng.gen::<f64>() * 0.04 - 0.02).collect();
        let predictions: Vec<f64> = actual_returns
            .iter()
            .map(|&r| r + rng.gen::<f64>() * 0.01 - 0.005)
            .collect();

        let strategy = TradingStrategy::new(0.001);
        let result = run_backtest(&predictions, &actual_returns, &strategy, 100000.0, 8760);

        assert_eq!(result.returns.len(), n);
        assert_eq!(result.equity_curve.len(), n);
        assert!(result.sharpe_ratio.is_finite());
        assert!(result.max_drawdown >= 0.0 && result.max_drawdown <= 1.0);
        assert!(result.win_rate >= 0.0 && result.win_rate <= 1.0);
    }

    #[test]
    fn test_buy_and_hold() {
        let returns = vec![0.01, -0.005, 0.02, -0.01, 0.015];
        let result = calculate_buy_and_hold(&returns, 100000.0, 252);

        assert_eq!(result.returns.len(), 5);
        assert_eq!(result.n_trades, 1);
    }

    #[test]
    fn test_compare_strategies() {
        use rand::{Rng, SeedableRng};
        use rand::rngs::StdRng;

        let mut rng = StdRng::seed_from_u64(42);
        let n = 100;

        let actual_returns: Vec<f64> = (0..n).map(|_| rng.gen::<f64>() * 0.04 - 0.02).collect();
        let predictions: Vec<f64> = actual_returns
            .iter()
            .map(|&r| r + rng.gen::<f64>() * 0.01 - 0.005)
            .collect();

        let thresholds = vec![0.0005, 0.001, 0.002];
        let results = compare_strategies(&predictions, &actual_returns, &thresholds, 8760);

        assert_eq!(results.len(), 3);
        for (threshold, result) in results {
            assert!(threshold > 0.0);
            assert!(result.sharpe_ratio.is_finite());
        }
    }
}
